[MoveLib] Initial checkin of the governance contract

lxfind opened the pull request:
This PR adds the initial governance contract in Move. It contains 4 components:
1. Genesis: This module creates a singleton system state object along with SUI treasury cap during init. All genesis init values are defined here.
2. SuiSystem: This module defines the system state type that contains all information of the system. Most entry functions that will be called from validators are here, such as add/withdraw validator, add/withdraw stake and etc.
3. ValidatorSet: This module defines the validator set. It doesn't contain any entry function, and is used exclusively by SuiSystem to maintain the validator set.
4. Validator: This module defines the Validator struct.

Workflows of different entities:
1. A non-validator

A non-validator can first call `Validator::create` to create a new Validator object with some initial stake. Given the object, they can also add more stake by calling `Validator::add_stake`, or withdraw stake with `Validator::withdraw_stake`, or destroy it and get back the stake by calling `Validator::destroy`.

Once they are satisfied with the object, they can call `SuiSystem::request_add_validator` to request to become a new validator. After the request is made, they no longer have access to the validator object and is now considered as a pending validator

2. A pending validator

Once someone did (1) and became a pending validator, they could continue adding stake to that pending validator by calling `SuiSystem::request_add_stake`, or withdraw some stake by calling `SuiSystem::request_withdraw_stake`. The effects of these two calls are immediate.
They can also call `SuiSystem::request_withdraw_validator`. The effect is also immediate: they will no longer be a pending validator and the object is sent back.

At the end of an epoch, if the pending validator meets all conditions, it will become an active validator.

3. An active validator.

Similarly, one can request adding/withdraw stake on an active validator. However the effect will not be immediate, and will only be processed at the end of epoch. They can also request to withdraw the validator itself, which will also be requested at the end of epoch.

TODO: Add unittests

Things not covered by this PR and will be done in future PRs:
1. gas pricing
2. stake award distribution
3. stake delegations

bmwill commented:
This is awesome, especially since there's been a number of questions recently about the shape of what the ValidatorSet would look like.

Just as a point of clarification, this doesn't currently support delegating stake to a Validator (either active or pending) correct? I suspect that would complicate things given you'd need extra bookkeeping for mapping delegated stake back to some other address not affiliated with the Validator.

bmwill commented on sui_programmability/framework/sources/Governance/Validator.move:
```diff
@@ -0,0 +1,212 @@
+        sui_address: address,
+        /// A unique human-readable name of this validator.
+        name: String,
+        /// The IP address of the validator.
+        ip_address: u128,
```
ip address may not be sufficient as we'll probably need to know port information. Also a Validator may wish to provide a DNS name so that it doesn't require a fixed static ip address.

lxfind replied:
Let me make this a vector<u8> so that it could contain anything

lxfind commented:
> Just as a point of clarification, this doesn't currently support delegating stake to a Validator (either active or pending) correct? I suspect that would complicate things given you'd need extra bookkeeping for mapping delegated stake back to some other address not affiliated with the Validator.

Right, I haven't implemented delegation yet. It will be quite complex.

sblackshear commented on sui_programmability/framework/sources/Governance/SuiSystem.move:
```diff
@@ -0,0 +1,162 @@
+    /// (i.e. sender must match the sui_address in the validator).
+    /// At the end of the epoch, the `validator` object will be returned to the sui_address
+    /// of the validator.
+    public(script) fun request_withdraw_validator(
```
Nit: I would call this something like `request_remove_validator`. "Withdraw" sounds like it might be about withdrawing stake rather than removing a validator from the validator set.

sblackshear commented on sui_programmability/framework/sources/Governance/ValidatorSet.move:
```diff
@@ -0,0 +1,296 @@
+    /// will be processed at the end of epoch.
+    public(friend) fun request_withdraw_validator(
+        self: &mut ValidatorSet,
+        validator_address: address,
```
Nit: I think it would be slightly safer/clearer to take `TxContext` as input + compute `sender` here instead of doing it in the caller. Similar elsewhere for `request_add_stake` etc.

sblackshear commented on sui_programmability/framework/sources/Governance/ValidatorSet.move:
```diff
@@ -0,0 +1,296 @@
+    ///   4. At the end, we decide which validators will form the validator set for the next epoch.
+    public(friend) fun advance_epoch(
+        self: &mut ValidatorSet,
+        new_validator_set_size: u64,
```
I think validator set size and min/max stake are not things that will change on an epoch-by-epoch basis. Any change would probably be part of a protocol upgrade + would (probably) take effect by publishing by updating `SystemParameters`?

sblackshear commented on sui_programmability/framework/sources/Governance/ValidatorSet.move:
```diff
@@ -0,0 +1,296 @@
+    /// Find validator by `validator_address`, in `validators`.
+    /// Returns (true, index) if the validator is found, and the index is its index in the list.
+    /// If not found, returns (false, 0).
+    fun find_validator(validators: &vector<Validator>, validator_address: address): (bool, u64) {
```
I think `Option<u64>` might be a bit safer, but not a big deal since this is an internal API.

sblackshear commented on sui_programmability/framework/sources/Governance/ValidatorSet.move:
```diff
@@ -0,0 +1,296 @@
+
+    const ENOT_ENOUGH_VALIDATORS: u64 = 4;
+
+    struct ValidatorSet has store {
```
I think the total stake and quorum threshold (i.e., 66% of the total stake) would also be useful to store here. Certificate checking logic will need to know these numbers in addition to the validator set.

sblackshear commented on sui_programmability/framework/sources/Governance/Validator.move:
```diff
@@ -0,0 +1,212 @@
+
+    /// Withdraw stake from a pending validator. Since it's pending, we can
+    /// withdraw coins immediately and send it back to the validator's address.
+    public(friend) fun withdraw_stake_from_pending_validator(
```
I think we should either not allow this (slightly preferred, since it's a bit simpler), or should only allow it if the amount of stake remaining is still above the minimum. Otherwise, someone can spam the pending validator set by adding themselves, then withdrawing all of their stake. We must try to ensure that the pending set doesn't get too big (e.g., hit the object size limit).

sblackshear commented on sui_programmability/framework/sources/Governance/ValidatorSet.move:
```diff
@@ -0,0 +1,296 @@
+            }
+        };
+        let length = Vector::length(validators);
+        // Note: This may cause epoch advancement to fail, if we don't
```
I think ideally, the only way this should fail is if there are 0 validators. We should have a min, but I don't think a max is needed.

lxfind commented on sui_programmability/framework/sources/Governance/Validator.move:
```diff
@@ -0,0 +1,212 @@
+        assert!(pending_withdraw == 0, EINCONSISTENT_STATE);
+        Transfer::transfer(stake, sender);
+        ID::delete(id);
+        Option::destroy_none(pending_stake);
```
Yes this is done through `check_non_active_validator_invariants` in a few places.

sblackshear replied:
Cool--might be useful to document for the paranoid reader.

lxfind commented:
Addressed all feedback:
1. ip_address is now vector<u8>
2. Added `max_validator_candidate_count` to constrain the max number of validator candidates
3. Do not allow stake withdraw or validator removal for pending validators (but still allow adding stake for pending validators)
4. We no longer sort the validators by stake, because any validator that meets the min/max stake requirement can become alive in the next epoch.
5. Add total_stake and quorum_threshold

gdanezis commented on sui_programmability/framework/sources/Governance/Genesis.move:
```diff
@@ -0,0 +1,56 @@
+    /// Basic information of Validator1, as an example, all dummy values.
+    const VALIDATOR1_SUI_ADDRESS: address = @0x1234;
+    const VALIDATOR1_NAME: vector<u8> = b"Validator1";
+    const VALIDATOR1_IP_ADDRESS: vector<u8> = x"00FF00FF";
```
Could we have URLs for this? (domain, port at least).

lxfind replied:
I leave it to the validator to decide how to interpret it (with `vector<u8>`). 
Or are you suggesting we should settle on what's included in the ip address field?

gdanezis replied:
Lets call it NET_ADDRESS and leave it open to interpretation for the moment.

lxfind commented:
Added tests

lxfind commented:
Any other feedback?

tnowacki commented:
> Any other feedback?

I'm reading it now. But mostly for my own edification so not sure how much feedback I'll have. Other than most of these files are missing a new line at the end :P (great feedback, I know)

tnowacki commented on sui_programmability/framework/sources/Governance/SuiSystem.move:
```diff
@@ -0,0 +1,173 @@
+        treasury_cap: TreasuryCap<SUI>,
+        storage_fund: Coin<SUI>,
+        max_validator_candidate_count: u64,
+        min_validator_stake: u64,
+        max_validator_stake: u64,
```
Should you assert that min <= max?

tnowacki commented on sui_programmability/framework/sources/Governance/Validator.move:
```diff
@@ -0,0 +1,234 @@
+    }
+
+    /// Called by `ValidatorSet`, to send back a Validator object to its address.
+    /// This happens when a validator is withdrawn or does not quality.
```
```suggestion
    /// This happens when a validator is withdrawn or does not qualify.
```
?

tnowacki commented on sui_programmability/framework/sources/Governance/ValidatorSet.move:
```diff
@@ -0,0 +1,299 @@
+            total_stake,
+            quorum_threshold_pct,
+            active_validators: init_active_validators,
+            pending_validators: Vector::empty(),
+            pending_removals: Vector::empty(),
```
Same as above, you can also use `vector[]`, but its whatever you like better

tnowacki replied:
Thats a bug, I'll look

tnowacki commented on sui_programmability/framework/sources/Governance/ValidatorSet.move:
```diff
@@ -0,0 +1,299 @@
+    fun process_pending_removals(validators: &mut vector<Validator>, withdraw_list: &mut vector<u64>) {
+        sort_removal_list(withdraw_list);
+        while (!Vector::is_empty(withdraw_list)) {
+            let index = Vector::pop_back(withdraw_list);
+            let validator = Vector::remove(validators, index);
```
This will be O(n^2), but not sure there is much else we can do about it now

lxfind replied:
hmm why is this O N^2?

tnowacki replied:
(assuming it is stable, which I think it is)

tnowacki replied:
Also is this even correct then? The indexes will all shift around after the first remove

lxfind replied:
`sort_removal_list` will sort the indexes, so we will be removing the highest index first, and etc.

tnowacki replied:
Ah, gotcha.... Still O(n^2) worst case, but should be a bit better in practice (probably roughly O(n), especially if n is smallish)

bmwill commented:
> Other than most of these files are missing a new line at the end :P (great feedback, I know)
@tnowacki I haven't gotten around to setting up robust linting like we have on the move repo yet. But I want to because this sort of stuff weirdly bothers me too lol

lxfind commented:
Addressed Alonso's feedback.

sblackshear commented on sui_programmability/framework/sources/Governance/ValidatorSet.move:
```diff
@@ -0,0 +1,284 @@
+        } = self;
+        while (!Vector::is_empty(&active_validators)) {
+            let v = Vector::pop_back(&mut active_validators);
+            // We don't care about the actural context,
```
```suggestion
            // We don't care about the actual context,
```

sblackshear reviewed (commented):
One thing I am wondering in reading through this code is whether things could become simpler if we took away `key` from `Validator`:
- This function would create a `Validator` + request addition in a single tx instead of requiring two
- `process_pending_removal`s could unpack the `Validator` and send its owner the stake
- The distinction between "active" and "inactive" validators (along with checks related to this) disappear--everything is either pending or active
- Some of the checks on the `Validator` structure performed here (e.g., sender check at 99) can be eliminated. Basically, these checks become data invariants on the `Validator` struct instead of invariants that hold only on active or pending `Validator`s.

This does mean that if you withdraw from the validator set or get kicked out due to insufficient stake, you'll have to pass your `name`, `net_address`, and `stake` back in if you want to rejoin. But I think this annoyance might be worth it for the simplicity it will buy, especially in the short term where we will likely need to be changing this code a lot, but won't have any annoyed validators.

lxfind commented:
@sblackshear Great question!
For what's written, yes using non-key Validator would be a lot simpler.
The only thing that I haven't made my mind around, is whether we ever want the Validator objects to be child objects, instead of wrapped ones.
There are some benefits there:
1. Keep the system state object small (esp if there is a size limit)
2. Easier for anyone to see the content of the individual validator object

lxfind commented:
Made `Validator` a non-key object
